8332259: JvmtiTrace::safe_get_thread_name fails if current thread is in native state

lmesnik opened the pull request:
The JvmtiTrace::safe_get_thread_name sometimes crashes when called while current thread is in native thread state.

It happens when thread_name is set for tracing from jvmti functions.
See:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/prims/jvmtiEnter.xsl#L649 

The setup is called and the thread name is used in tracing before the thread transition. There is no good location where this method could be called from vm thread_state only. Some functions like raw monitor enter/exit never transition in vm state. So sometimes it is needed to call this function from native thread state.

The change should affect JVMTI trace mode only (-XX:TraceJVMTI). 

Verified by running jvmti/jdi/jdb tests with tracing enabled.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332259](https://bugs.openjdk.org/browse/JDK-8332259): JvmtiTrace::safe_get_thread_name fails if current thread is in native state (**Bug** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [12ddfca2](https://git.openjdk.org/jdk/pull/19275/files/12ddfca2c1e316184ba9ec58935529914e99ef7e)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19275/head:pull/19275` \
`$ git checkout pull/19275`

Update a local copy of the PR: \
`$ git checkout pull/19275` \
`$ git pull https://git.openjdk.org/jdk.git pull/19275/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19275`

View PR using the GUI difftool: \
`$ git pr show -t 19275`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19275.diff">https://git.openjdk.org/jdk/pull/19275.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19275#issuecomment-2116486952)

dholmes-ora reviewed (commented):
I have to wonder whether this solution will potentially cause problems because the code will now block for safepoints. We could fallback to `Thread::name()` if the current thread is in-native.

lmesnik commented:
> I have to wonder whether this solution will potentially cause problems because the code will now block for safepoints. We could fallback to `Thread::name()` if the current thread is in-native.

Thanks for feedback. Here is the update.

I've updated the safe_get_thread_name() to not change thread state. 

In  "jvmtiEnter.xsl" functions the thread name is s read once before the transition happened and re-used then. 
So I updated the tracing to 're-read' if the transition to VM happened to update the thread name once it became known.

dholmes-ora commented:
I don't understand the additional changes because they read the current thread's name, whereas this issue is about reading an arbitrary thread's name when the current thread happens to be in the wrong state. ???

dholmes-ora commented on src/hotspot/share/prims/jvmtiEventController.cpp:
```diff
@@ -958,7 +958,6 @@ JvmtiEventControllerPrivate::clear_to_frame_pop(JvmtiEnvThreadState *ets, JvmtiF
 void
 JvmtiEventControllerPrivate::change_field_watch(jvmtiEvent event_type, bool added) {
   int *count_addr;
-
```
Nit: this file doesn't need to be touched.

dholmes-ora reviewed (approved):
Okay now I get it. Even once the function is made truly safe, we are always calling it from an unsafe state and so will get the default `Thread::name` response. So now, after any transition to the VM the name is read again to get a good value. This seems a good enhancement though I have to wonder if the apparent changing of the thread name in the tracing might cause problems. The tracing really needs to include a unique thread identifier.

Thanks

AlanBateman commented:
Are there tests that run with TraceJVMTI so that this option is tested?

lmesnik commented:
There are no tests currently executed with TraceJVMTI.
I am thinking about adding execution of the svc testing. However I've got another failure that should resolved before
https://bugs.openjdk.org/browse/JDK-8332536
It is not related to the current issue.
Probably, it makes sense to add some basic logging testing with verification of log content also.

lmesnik commented:
I discussed the issue with @sspitsyn and after this, I updated the function to return something more descriptive in case we can't read thread state. (Easier to understand what happens.)

sspitsyn commented on src/hotspot/share/prims/jvmtiTrace.cpp:
```diff
@@ -277,6 +278,12 @@ const char *JvmtiTrace::safe_get_thread_name(Thread *thread) {
+  if (Thread::current()->is_Java_thread()) {
+    JavaThreadState current_state = JavaThread::cast(Thread::current())->thread_state();
+    if (current_state == _thread_in_native || current_state == _thread_blocked) {
+      return "not readable";
```
Nit: I'd suggest to make it more detailed, something like like this:
 "<non-readable thread name>" or "<NON-READABLE-THREAD-NAME>"

dholmes-ora replied:
Yes this would have looked better if the text was more clearly an error message with angle brackets.

lmesnik replied:
@sspitsyn, @dholmes-ora  Thanks for the naming suggestion, looks to long in the report. Let me try to use logging and see if it makes sense to make more improvements.

sspitsyn reviewed (approved):
This looks good, Posted one nit though.

lmesnik commented:
The name becomes too long in the logs.

lmesnik commented:
/integrate
